### ISSUE TITLE
Send special keys

### ISSUE THREAD

**Issue by MarSoft** [piater/uinputchars#1]
Is it possible to send some hotkey combination with this tool? Something like Control+Meta+X

**Comment by piater** [comment]
No, but it can be easily implemented.

Your question already identifies the core issue :-) The output generation should be trivially extensible to all relevant modifier keys. For input, one possibility is to define an escape syntax for modifier and other special keys. However, I feel that this violates the spirit of this tool that essentially implements an inverse keymap. I'd prefer to support hotkey combinations by special command-line options. For example, `echo a | uinputchars -C` might emit a Control+a.

I'll consider patches or may even implement it myself someday if people want it.

**Comment by MarSoft** [comment]
What about stuff like Ctrl+F4? They don't have corresponding text for stdin
at all. Or maybe pass them as modifiers? Like this: echo -n | uinputchars
-m Ctrl+F4

вт, 18 февр. 2020 г., 0:41 Justus Piater <notifications@github.com>:

**Comment by piater** [comment]
Good point. Technically, function keys are also subject to keymap translation, as are modifier keys. So maybe the best course of action is an escape syntax after all. Since key press and key release events are emitted separately, one might separate the issues of literal vs. escaped input and key press vs release events. Thinking aloud, how about:

- `\`: escape character
- `\\`: literal backslash
- `\(`: next token represents a key press event
- `\)`: key pressed at corresponding `\(` is released
- `\K_...`: action code to be translated by inverse keymap to corresponding keycode `KEY_...` to be emitted
- `\KEY_...`: keycode to be emitted, bypassing the inverse keymap
- ` `: terminates `\K...`  tokens; can be omitted if next token is also escaped
- any other character: character to be translated by inverse keymap
- any token (other than `\(` and `\)`) not immediately preceded by `\(` results in a key press and a key release event in immediate succession

Examples:

- Ctrl+F4: `\(\K_CTRL\K_F4\)`
- The following result in the same emitted sequence: `A`, `\(\K_SHIFTL a\)`, `\(\K_SHIFTL\(a\)\)`
- To save the current file in emacs: `\(\K_CTRL xs\)`
- To switch to VT 2: `\(\K_CTRL\(\K_ALT\K_F2\)\)`
